spark3/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java

```diff
@@ -0,0 +1,232 @@
+        table.updateProperties()
+            .set(TableProperties.DEFAULT_NAME_MAPPING, mappingJson)
+            .commit();
+      }
+
+      if (isFile || !partition.isEmpty()) {
+        // we do a list operation on the driver to import 1 file or 1 partition
+        PartitionSpec spec = table.spec();
+        MetricsConfig metricsConfig = MetricsConfig.fromProperties(table.properties());
+        String partitionURI;
+
+        if (isFile) {
+          partitionURI = dataPath.toString();
+        } else {
+          partitionURI = dataPath.toString() + toPartitionPath(partition);
+        }
+
+        List<DataFile> files = SparkTableUtil.listPartition(partition, partitionURI, format, spec, conf, metricsConfig);
+        if (files.size() == 0) {
+          throw new IllegalArgumentException(String.format("No files found for add_file command. Looking in URI %s",
+              partitionURI));
+        }
+
+        AppendFiles append = table.newAppend();
+        files.forEach(append::appendFile);
+        append.commit();
+      } else {
+        // Importing multiple partitions
+        List<SparkTableUtil.SparkPartition> partitions = Spark3Util.getPartitions(spark(), dataPath, format);
+        String stagingLocation = table.properties()
+            .getOrDefault(TableProperties.WRITE_METADATA_LOCATION, table.location() + "/metadata");
+        SparkTableUtil.importSparkPartitions(spark(), partitions, table, table.spec(), stagingLocation);
+      }
+
+      Snapshot snapshot = table.currentSnapshot();
+      Long numAddedFiles = Long.parseLong(snapshot.summary().get(SnapshotSummary.ADDED_FILES_PROP));
+      return numAddedFiles;
+    });
+  }
+
+  @Override
+  public String description() {
+    return null;
+  }
+
+  private String toPartitionPath(Map<String, String> partition) {
+    return partition.entrySet().stream()
+        .map(entry -> entry.getKey() + "=" + entry.getValue())
+        .collect(Collectors.joining("/", "/", ""));
+  }
+
+  private void validatePartitionSpec(Table table, Path path, FileSystem fs, Map<String, String> partition) {
+
+    List<PartitionField> partitionFields = table.spec().fields();
+
+    boolean tablePartitioned = !partitionFields.isEmpty();
+    boolean partitionSpecPassed = !partition.isEmpty();
+
+    if (tablePartitioned && partitionSpecPassed) {
+      // Has Partitions, Check that they are valid
+      if (partitionFields.size() != partition.size()) {
+        throw new IllegalArgumentException(
+          String.format(
+            "Cannot add data files to target table %s because that table is partitioned, " +
+            "but the number of columns in the provided partition spec (%d) " +
+            "does not match the number of partitioned columns in table (%d)",
+              table.name(), partition.size(), partitionFields.size()));
+      }
+      partitionFields.forEach(field -> {
+        if (!partition.containsKey(field.name())) {
```

Comment:
From the above discussion, I gather that Hive partitions can also be imported? Do we see issues here around Hive's lowercasing of columns? If yes, then something to consider for future enhancements...

Reply:
Yeah this could be a problem if the entries that Spark returns are different than those we expect based on column names, i'll need to add some hive tests for that